qiskit/transpiler/passes/synthesis/hls_plugins.py

```diff
@@ -1803,67 +1813,110 @@ def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **
-            # - main_tracker.set_dirty(base_qubits)
-            #
-            # Note that we need to set the base_qubits to dirty if we have power or inverse
-            # modifiers. For power: even if the power is a positive integer (that is, we need
-            # to repeat the same circuit multiple times), even if the target is initially at |0>,
-            # it will generally not be at |0> after one iteration. For inverse: as we
-            # flip the order of operations, we cannot exploit which qubits are at |0> as "viewed from
-            # the back of the circuit". If we just have control modifiers, we can use the state
-            # of base qubits when synthesizing the controlled operation.
-            #
-            # In addition, all of the other global qubits that are not a part of the annotated
-            # operation can be used as they are in all cases, since we are assuming that all of
-            # the synthesis methods preserve the states of ancilla qubits.
-            annotated_tracker = tracker.copy()
-            control_qubits = input_qubits[:num_ctrl]
-            base_qubits = input_qubits[num_ctrl:]
-            annotated_tracker.disable(control_qubits)  # do not access control qubits
-            if power != 0 or is_inverted:
-                annotated_tracker.set_dirty(base_qubits)
-
-            # Note that synthesize_operation also returns the output qubits on which the
-            # operation is defined, however currently the plugin mechanism has no way
-            # to return these (and instead the upstream code greedily grabs some ancilla
-            # qubits from the circuit). We should refactor the plugin "run" iterface to
-            # return the actual ancilla qubits used.
-            synthesized_base_op_result = synthesize_operation(
-                operation.base_op, input_qubits[num_ctrl:], data, annotated_tracker
-            )
+        # The synthesis consists of two steps:
+        #   - First, we synthesize the base operation.
+        #   - Second, we apply modifiers to this circuit.
+        #
+        # An important optimization (similar to the code in ``add_control.py``) is to synthesize
+        # the base operation with respect to a larger set of "basis" gates, to which the control
+        # logic can be added more efficiently. In addition, we add annotated operations to be
+        # in this larger set, exploiting the fact that adding control to annotated operations
+        # returns a new annotated operation with an extended list of modifiers.
+        #
+        # Note that it is fine for this function to return a circuit with high-level objects
+        # (including annotated operations) as the HighLevelSynthesis transpiler pass will
+        # recursively re-synthesize this circuit, However, we should always guarantee that some
+        # progress is made.
+        basis = set(EFFICIENTLY_CONTROLLED_GATES + ["annotated", "mcx", "qft"])
```

Comment:
are these the only relevant gates? how about mcphase? or mcrz/mcry/mcrx?

Reply:
This is also a great question. For ``MCPhaseGate``, we don't have a plugin mechanism (correct?), so the definition circuit will be used. Can we do something better using ancillas? Is an ``MCPhaseGate`` used when constructing more complex circuits? If yes, then it could make sense to first unroll a more complex circuit to a circuit with ``MCPhaseGate``, then apply the best known algorithm for ``MCPhaseGate``. There is also a question on whether we need a native ``MCPhaseGate `` at all, as opposed to a controlled phase gate. Bottom-line: I agree that we should think about the above, but I would not like to add more gates here without concrete use cases.

Reply:
what about adders? maybe in the future? say someone wants to add a controlled adder (w/o specifying the method)